### PR TITLE
[wmco] Add WMCO version to Windows node

### PR DESCRIPTION
**Pull request by sebsoto**
This commit adds the version of the WMCO instance that configured each
node to that node's annotations. This will be used in the WMCO upgrade
process, as this allows WMCO to see which nodes were not configured by
the current version.

### CODE REVIEW

File: test/e2e/validation_test.go [hunk 1]
```diff
@@ -43,6 +44,14 @@ func testWorkerLabel(t *testing.T) {
 	}
 }
 
+// testVersionAnnotation tests all nodes are annotated with the version of the currently deployed WMCO
+func testVersionAnnotation(t *testing.T) {
+	for _, node := range gc.nodes {
+		require.Containsf(t, node.Annotations, nc.VersionAnnotation, "node %s missing version annotation", node.Name)
```
Comment (VaishnaviHire) [inline test/e2e/validation_test.go:50]: nit: `node.GetName()`

File: pkg/controller/windowsmachine/nodeconfig/nodeconfig.go [hunk 2]
```diff
@@ -178,13 +181,14 @@ func (nc *nodeConfig) configureNetwork() error {
...
+		nc.node.Labels[WorkerLabel] = ""
+	} else {
 		log.V(1).Info("worker label %s already present on node %s", WorkerLabel, nc.node.GetName())
-		return nil
 	}
-	nc.node.Labels[WorkerLabel] = ""
+	nc.node.Annotations[VersionAnnotation] = version.Get()
 	node, err := nc.k8sclientset.CoreV1().Nodes().Update(context.TODO(), nc.node, metav1.UpdateOptions{})
 	if err != nil {
 		return errors.Wrap(err, "error applying worker label to node object")
```
Comment (mansikulkarni96) [inline pkg/controller/windowsmachine/nodeconfig/nodeconfig.go:194]: error message needs to be updated for version annotation

File: hack/common.sh [hunk 3]
```diff
@@ -91,7 +91,7 @@ cleanup_WMCO() {
 # returns the operator version in `Version+GitHash` format
 get_version() {
   OPERATOR_VERSION=0.0.1
-  GIT_COMMIT=$(git rev-parse --short HEAD)
+  GIT_COMMIT=$(git rev-parse HEAD)
```
Comment (ravisantoshgudimetla) [inline hack/common.sh:94]: Why this change?

File: test/e2e/wmco_test.go [hunk 4]
```diff
@@ -48,3 +53,18 @@ func setupWMCOResources() error {
 	}
 	return nil
 }
+
+// wmcoVersion returns the version that should be annotated on each Windows node
+func wmcoVersion() (string, error) {
```
Comment (ravisantoshgudimetla) [inline test/e2e/wmco_test.go:58]: nit: I'd have injected this value into the custom struct by calling `wmco version` command or something like that.

File: cmd/manager/main.go [hunk 5]
```diff
@@ -152,8 +152,8 @@ func main() {
 		os.Exit(1)
 	}
 
-	// watch `openshift-machine-api` namespace along with wmco namespace
-	namespaces := []string{"openshift-machine-api", namespace}
+	// watch `""` (nodes), `openshift-machine-api` (machines), namespaces, along with the wmco namespace
```
Comment (aravindhp) [inline cmd/manager/main.go:155]: It is unclear what `""` implies. Please make it `WMCO` given you are editing this line.

File: hack/common.sh (hunk 3 above)
Comment (aravindhp) [reply hack/common.sh:94]: Is this because of `The version format was changed to full hash instead of shortened hash to
remove possibility of collisions and allow for better validation.`? I am -1 on having the full hash added to the version. We get reasonable collision prevnetion and validation with the short hash and for version that is good enough.

File: pkg/controller/add_node.go [hunk 6]
```diff
@@ -0,0 +1,10 @@
+package controller
```
Comment (aravindhp) [inline pkg/controller/add_node.go:1]: Did you add this manually or run some SDK command?

File: pkg/controller/node/node_controller.go [hunk 7]
```diff
@@ -0,0 +1,127 @@
+package node
+
+import (
+	"context"
+
+	"github.com/openshift/windows-machine-config-operator/pkg/controller/windowsmachine/nodeconfig"
+	"github.com/openshift/windows-machine-config-operator/version"
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
```
Comment (aravindhp) [inline pkg/controller/node/node_controller.go:9]: `core` is sufficient.

File: pkg/controller/node/node_controller.go [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+			return false
+		},
+		GenericFunc: func(e event.GenericEvent) bool {
+			return false
+		},
+	}
+
+	// Watch for changes to primary resource Node
+	err = c.Watch(&source.Kind{Type: &corev1.Node{}}, &handler.EnqueueRequestForObject{}, predicateFilter)
+	if err != nil {
```
Comment (aravindhp) [inline pkg/controller/node/node_controller.go:78]: One line please

File: pkg/controller/node/node_controller.go [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+
+// Add creates a new Node Controller and adds it to the Manager. The Manager will set fields on the Controller
+// and Start it when the Manager is Started.
+func Add(mgr manager.Manager, _ string) error {
+	return add(mgr, newReconciler(mgr))
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager) reconcile.Reconciler {
+	return &ReconcileNode{client: mgr.GetClient(), scheme: mgr.GetScheme()}
```
Comment (aravindhp) [inline pkg/controller/node/node_controller.go:33]: Are you sure we don't need the non-cached client?

File: pkg/controller/node/node_controller.go [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+
+// ReconcileNode reconciles a Node object
+type ReconcileNode struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client client.Client
+	scheme *runtime.Scheme
+}
+
+// Reconcile reads that state of the cluster for a Node object and makes changes based on the state read
```
Comment (aravindhp) [inline pkg/controller/node/node_controller.go:96]: I would just say `state of a Node object`

File: pkg/controller/node/node_controller.go [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	// Set version annotation to the proper value
+	instance.Annotations[nodeconfig.VersionAnnotation] = version.Get()
+	err = r.client.Update(context.TODO(), instance, &client.UpdateOptions{})
+	if err != nil {
```
Comment (aravindhp) [inline pkg/controller/node/node_controller.go:121]: Shouldn't this update be retried on failure?

File: pkg/controller/windowsmachine/nodeconfig/nodeconfig.go [hunk 12]
```diff
@@ -30,6 +31,8 @@ const (
 	WindowsOSLabel = "node.openshift.io/os_id=Windows"
 	// WorkerLabel is the label that needs to be applied to the Windows node to make it worker node
 	WorkerLabel = "node-role.kubernetes.io/worker"
+	// VersionAnnotation indicates the version of WMCO that configured the node
+	VersionAnnotation = "windowsmachineconfiguration.openshift.io/config-version"
```
Comment (aravindhp) [inline pkg/controller/windowsmachine/nodeconfig/nodeconfig.go:35]: I am not so sure if `windowmachineconfiguration` is the right choice. Is there a reason for `config-version` vs `version`? I prefer the shorter `version`. Plus there is some stuttering going on. How about `windowsmachineconfig.openshift.io/version`?

File: pkg/controller/windowsmachine/nodeconfig/nodeconfig.go [hunk 13]
```diff
@@ -131,9 +134,8 @@ func (nc *nodeConfig) Configure() error {
 	if err := nc.setNode(); err != nil {
 		return errors.Wrapf(err, "error getting node object for VM %s", nc.ID())
 	}
-	// Apply worker labels
-	if err := nc.applyWorkerLabel(); err != nil {
-		return errors.Wrap(err, "failed applying worker label")
+	if err := nc.applyLabelsAndAnnotations(); err != nil {
```
Comment (aravindhp) [inline pkg/controller/windowsmachine/nodeconfig/nodeconfig.go:137]: Please have separate functions for labels and annotations. The `Update` can be moved out. This will give us more flexibility in the future.

File: test/e2e/validation_test.go [hunk 14]
```diff
@@ -43,6 +44,40 @@ func testWorkerLabel(t *testing.T) {
...
+		require.NoError(t, err, "error updating node objects")
+	}
+
+	// repopulate gc.nodes with new values
+	nodes, err := tc.kubeclient.CoreV1().Nodes().List(context.TODO(), metav1.ListOptions{LabelSelector: nc.WindowsOSLabel})
+	require.NoError(t, err, "error listing Windows nodes")
+	require.Lenf(t, nodes.Items, int(gc.numberOfNodes), "Expected num of Windows nodes does not match: %v", nodes.Items)
+	gc.nodes = nodes.Items
+	for _, node := range gc.nodes {
+		require.Containsf(t, node.Annotations, nc.VersionAnnotation, "node %s missing version annotation", node.GetName())
```
Comment (aravindhp) [inline test/e2e/validation_test.go:75]: It will be a good data point for us if one node gets updated and not the other. So the loop should not stop on the first failure.

File: test/e2e/wmco_test.go (hunk 4 above)
Comment (aravindhp) [reply test/e2e/wmco_test.go:58]: I would rather we use `wmco version`. You can further validate by comparing with `git` if you want but `wmco version` should be the source of truth.

File: hack/common.sh (hunk 3 above)
Comment (sebsoto) [reply hack/common.sh:94]: from the commit message:

"The version format was changed to full hash instead of shortened hash to
remove possibility of collisions and allow for better validation."

File: cmd/manager/main.go [hunk 15]
```diff
@@ -152,8 +152,8 @@ func main() {
...
+	namespaces := []string{"", "openshift-machine-api", namespace}
```
[the remark this answers is not in the document]
Comment (sebsoto) [reply cmd/manager/main.go:156]: Updated comment to make that more clear

File: pkg/controller/add_node.go (hunk 6 above)
Comment (sebsoto) [reply pkg/controller/add_node.go:1]: I used the sdk command to add a controller

**Comment by sebsoto** [comment]
Noticed a weirdness happening in the test, Windows node failed to be found but bootstrapping reported completion and must-gather had the node's kubelet logs. Retested to check if this is a flake, and if so we may need to increase the node list timeout

File: README.md [hunk 16]
```diff
@@ -61,15 +61,19 @@ export OPERATOR_IMAGE=<registry url for remote WMCO image>
...
 ```
-We assume that the developer uses `openshift-dev` as the key pair in the aws cloud
+We assume that the developer uses `openshift-dev` as the key pair in the aws cloud.
 
 Additional flags that can be passed to `hack/run-ci-e2e-test.sh` are
 - `-s` to skip the deletion of Windows nodes that are created as part of test suite run
 - `-n` to represent the number of Windows nodes to be created for test run
 - `-k` to represent the AWS specific key pair that will be used during e2e run and it should map to the private key
        that we have in `KUBE_SSH_KEY_PATH`. The default value points to `openshift-dev` which we use in our CI
+- `-b` gives an alternative path to the WMCO binary. If the -b option is not given outside of CI, certain validation
```
Comment (aravindhp) [inline README.md:73]: - Is this is a required parameter when running locally? If yes, then it should be called out.
- Ideally it would be nice for this to be an option parameter with a sensible default.
- What validations tests fail? It would help to call that out here.

File: test/e2e/validation_test.go [hunk 17]
```diff
@@ -43,6 +45,35 @@ func testWorkerLabel(t *testing.T) {
 	}
 }
 
+// wmcoVersion returns the version of the operator. This is sourced from the WMCO binary used to create the operator image.
+// This function will return an error if the binary is missing.
+func wmcoVersion() (string, error) {
```
Comment (aravindhp) [inline test/e2e/validation_test.go:50]: Given this is not associated with a struct, getWMCOVersion() would be more appropriate.

File: pkg/controller/windowsmachine/nodeconfig/nodeconfig.go [hunk 18]
```diff
@@ -129,10 +132,16 @@ func (nc *nodeConfig) Configure() error {
...
 	}
-	// Apply worker labels
-	if err := nc.applyWorkerLabel(); err != nil {
-		return errors.Wrap(err, "failed applying worker label")
+	// Apply labels and annotations to nc.node
+	nc.addWorkerLabel()
+	nc.addVersionAnnotation()
+	// update node object
+	node, err := nc.k8sclientset.CoreV1().Nodes().Update(context.TODO(), nc.node, metav1.UpdateOptions{})
+	if err != nil {
```
Comment (aravindhp) [inline pkg/controller/windowsmachine/nodeconfig/nodeconfig.go:140]: Should we fail immediately or retry a few times? It is very possible that a change could be in flight when this update is made.

File: README.md (hunk 16 above)
Comment (sebsoto) [reply README.md]: Changed this to use the normal dev workflow as default.

File: pkg/controller/windowsmachine/nodeconfig/nodeconfig.go (hunk 18 above)
Comment (sebsoto) [reply pkg/controller/windowsmachine/nodeconfig/nodeconfig.go:140]: Its most likely safer to just requeue on error

File: README.md [hunk 19]
```diff
@@ -63,13 +63,17 @@ Once the above variables are set, run the following script:
...
 ```
-We assume that the developer uses `openshift-dev` as the key pair in the aws cloud
+We assume that the developer uses `openshift-dev` as the key pair in the aws cloud.
 
 Additional flags that can be passed to `hack/run-ci-e2e-test.sh` are
 - `-s` to skip the deletion of Windows nodes that are created as part of test suite run
 - `-n` to represent the number of Windows nodes to be created for test run
 - `-k` to represent the AWS specific key pair that will be used during e2e run and it should map to the private key
        that we have in `KUBE_SSH_KEY_PATH`. The default value points to `openshift-dev` which we use in our CI
+- `-b` gives an alternative path to the WMCO binary. This option is not acknowledged when running in CI.
```
Comment (aravindhp) [inline README.md:73]: Nit: I don't follow what you mean by `This option is not acknowledged when running in CI.`.

File: README.md [hunk 20]
```diff
@@ -63,13 +63,17 @@ Once the above variables are set, run the following script:
...
+       When building the operator locally, the WMCO binary is created as part of the operator image build process and
+       can be found at `build/_output/bin/windows-machine-config-operator`, this is the default value of this option.
```
Comment (aravindhp) [inline README.md:75]: Where is this default being set?

File: test/e2e/validation_test.go [hunk 21]
```diff
@@ -43,6 +45,35 @@ func testWorkerLabel(t *testing.T) {
...
+	return versionSplit[1], nil
+}
+
+// testVersionAnnotation tests all nodes are annotated with the version of the currently deployed WMCO
+func testVersionAnnotation(t *testing.T) {
+	operatorVersion, err := getWMCOVersion()
+	require.NoError(t, err, "could not get WMCO version")
+	for _, node := range gc.nodes {
+		t.Run(node.GetName(), func(t *testing.T) {
+			require.Containsf(t, node.Annotations, nc.VersionAnnotation, "node %s missing version annotation", node.GetName())
```
Comment (aravindhp) [inline test/e2e/validation_test.go:71]: I thought I mentioned this before. We should not stop the test on the first failure. It will be a good data point for us if there is a scenario where one node has the correct version annotation while the other does not.

File: pkg/controller/add_node.go (hunk 6 above)
Comment (aravindhp) [reply pkg/controller/add_node.go]: Please add the SDK command to your commit message.

File: test/e2e/validation_test.go (hunk 21 above)
Comment (sebsoto) [reply test/e2e/validation_test.go:71]: We are not stopping at the first failure here, each node is checked within a separate subtest https://github.com/openshift/windows-machine-config-operator/pull/88/files#diff-1a876af15bc4507294c2bc7e5f81dc37R70 

When one node fails, its subtest fails and the rest of the nodes are tested with their subtests.

File: README.md (hunk 20 above)
Comment (sebsoto) [reply README.md:75]: Being set here: https://github.com/openshift/windows-machine-config-operator/pull/88/files#diff-e8664074431f2653c7da0fc9019a1935R95

Comment (aravindhp) [reply README.md:75]: Sorry I missed that.

File: test/e2e/validation_test.go (hunk 21 above)
Comment (aravindhp) [reply test/e2e/validation_test.go:71]: You are right. I missed this too.

File: README.md (hunk 19 above)
Comment (sebsoto) [reply README.md]: Reworded

File: README.md [hunk 22]
```diff
@@ -63,13 +63,17 @@ Once the above variables are set, run the following script:
...
 ```
-We assume that the developer uses `openshift-dev` as the key pair in the aws cloud
+We assume that the developer uses `openshift-dev` as the key pair in the aws cloud.
 
 Additional flags that can be passed to `hack/run-ci-e2e-test.sh` are
 - `-s` to skip the deletion of Windows nodes that are created as part of test suite run
 - `-n` to represent the number of Windows nodes to be created for test run
 - `-k` to represent the AWS specific key pair that will be used during e2e run and it should map to the private key
        that we have in `KUBE_SSH_KEY_PATH`. The default value points to `openshift-dev` which we use in our CI
+- `-b` gives an alternative path to the WMCO binary. This option overrided in OpenShift CI.
```
Comment (aravindhp) [inline README.md:73]: `is overridden in...`

**Comment by VaishnaviHire** [comment]
LGTM
